src/Components/Web.JS/src/Services/NavigationEnhancement.ts

```diff
@@ -148,6 +149,15 @@ export async function performEnhancedPageLoad(internalDestinationHref: string, f
   // Now request the new page via fetch, and a special header that tells the server we want it to inject
   // framing boundaries to distinguish the initial document and each subsequent streaming SSR update.
   currentEnhancedNavigationAbortController = new AbortController();
+  if (currentUrl) {
+    currentEnhancedNavigationAbortController.signal.addEventListener('abort', () => {
+      // If the navigation was aborted, we need to revert the URL change
+      if (location.href === internalDestinationHref && location.href !== currentUrl) {
+        history.replaceState(null, /* ignored title */ '', currentUrl);
+      }
+    }, { once: true });
+  }
+
```

Comment:
The rationale behind this is:
* Capture the URL before it gets changed by clicking a link.
* Pass it down to the code that performs the actual page load.
* If the load is cancelled, then another enhanced nav has replaced this one.
* Check if the thing that triggered the cancellation already updated the URL
* If it didn't, and the current url is different from the previous url, then re-apply the old url.

I think this might still not handle forms that explicitly target a different endpoint, but I haven't tried that yet. We can probably tweak the behavior based on the presence of fetchOptions

Reply:
Unfortunately I don't think this covers it. Various things can still go wrong, e.g.:

 * Enhanced navigation isn't always triggered by a link click. It can also be triggered by back/forwards, at which point the "current URL" has already changed before you got a chance to capture it. Or it could be triggered by a different form submission.
 * Replacing the URL via `replaceState` is going to corrupt the history stack even further. This code doesn't know how many times the user did back/forwards, so there's no way to say if replacing the current history entry will produce the desired result or will just create further weirdness if you do a later back/forwards

Let's discuss this offline. Hopefully we can find a different way of constraining the problem so that there aren't so many edge cases.